data-generation/generate-instructions.py

```diff
@@ -0,0 +1,204 @@
+from random import randint, random
+
+
+"""
+Classes
+"""
+
+
+class NoActionException(Exception):
+    pass
+
+
+class Block:
+    def __init__(self, side1_letter,
+                 side1_color,
+                 side2_letter,
+                 side2_color,
+                 pos):
+        self.side1_letter = side1_letter
+        self.side1_color = side1_color
+        self.side2_letter = side2_letter
+        self.side2_color = side2_color
+        self.position = pos
+
+    def __eq__(self, other):
+        return self.side1_letter == other.side1_letter and \
+            self.side2_letter == other.side2_letter    and \
+            self.side1_color == other.side1_color      and \
+            self.side2_color == other.side2_color      or  \
+                                                           \
+            self.side1_letter == other.side2_letter    and \
+            self.side2_letter == other.side1_letter    and \
+            self.side1_color == other.side2_color      and \
+            self.side2_color == other.side1_color
+
+    def __ne__(self, other):
+        return not self == other
+
+    def shift_to(self, position):
+        return Block(self.side1_letter,
+                     self.side1_color,
+                     self.side2_letter,
+                     self.side2_color,
+                     position)
+
+    def flip(self):
+        return Block(self.side2_letter,
+                     self.side2_color,
+                     self.side1_letter,
+                     self.side1_color,
+                     self.position)
+
+    def __str__(self):
+        return '{side1: (%s, %s), side2: (%s, %s), pos: %s}' % \
+            (self.side1_letter, self.side1_color,
+             self.side2_letter, self.side2_color,
+             self.position)
+
+
+class Instruction:
+    def __init__(self, phrase, point):
+        self.phrase = phrase
+        self.point = point
+
+    def __str__(self):
+        return self.phrase
+
+
+class Configuration:
```

Comment:
Similar to `Block`, I see you're going for an immutable `Configuration`. Again, what do we gain?

Reply:
See above comment